File: benchmarks/cases/quicksort/quicksort.go
package main

import "fmt"

func quicksort(a []int64, lo, hi int64) {
	if lo >= hi {
		return
	}
	pivot := a[hi]
	i := lo - 1
	for j := lo; j < hi; j++ {
		if a[j] < pivot {
			i++
			a[i], a[j] = a[j], a[i]
		}
	}
	p := i + 1
	a[p], a[hi] = a[hi], a[p]
	quicksort(a, lo, p-1)
	quicksort(a, p+1, hi)
}

func main() {
	var n int64 = 1000000
	a := make([]int64, n)
	for i := int64(0); i < n; i++ {
		a[i] = (i*1103515245 + 12345) % 1000000
	}

	quicksort(a, 0, n-1)

	var chk int64 = 0
	for i := int64(0); i < n; i++ {
		chk += a[i] * i
	}
	fmt.Println(chk)
}

File: benchmarks/cases/quicksort/quicksort.cpp
#include <cstdio>

void quicksort(long* a, long lo, long hi) {
    if (lo >= hi) { return; }
    long pivot = a[hi];
    long i = lo - 1;
    for (long j = lo; j < hi; j++) {
        if (a[j] < pivot) {
            i++;
            long t = a[i];
            a[i] = a[j];
            a[j] = t;
        }
    }
    long p = i + 1;
    long t = a[p];
    a[p] = a[hi];
    a[hi] = t;
    quicksort(a, lo, p - 1);
    quicksort(a, p + 1, hi);
}

int main() {
    long n = 1000000;
    long* a = new long[n];
    for (long i = 0; i < n; i++) {
        a[i] = (i * 1103515245 + 12345) % 1000000;
    }

    quicksort(a, 0, n - 1);

    long chk = 0;
    for (long i = 0; i < n; i++) {
        chk = chk + a[i] * i;
    }
    printf("%ld\n", chk);

    delete[] a;
    return 0;
}
